variantlib/commands/generate_index_json.py

```diff
@@ -126,13 +100,39 @@ def generate_index_json(args: list[str]) -> None:
                 )
                 continue
 
-            if vdesc.hexdigest not in known_variants:
-                known_variants[variant_hash] = vdesc
+            if (vhash := vdesc.hexdigest) not in known_variants:
+                known_variants[vhash] = vdesc
+
+            # ============== Variant Providers Processing ================ #
+            if not (
+                wheel_providers := wheel_metadata.get_all(
+                    METADATA_VARIANT_PROVIDER_HEADER, []
+                )
+            ):
+                logger.info(
+                    "%(wheel)s: did not declare any `%(key)s`",
+                    {"wheel": wheel, "key": METADATA_VARIANT_PROVIDER_HEADER},
+                )
+
+            for wheel_provider in wheel_providers:
+                ns, _, provider = wheel_provider.partition(", ")
+                ns = ns.strip()
+                provider = provider.strip()
+                if not ns or not provider:
+                    raise ValueError(
+                        f"{wheel}: Invalid {METADATA_VARIANT_PROVIDER_HEADER}: "
+                        f"{wheel_provider}"
+                    )
+                providers[ns].add(provider)
+
+    sorted_providers = {}
+    for key, values in providers.items():
+        sorted_providers[key] = sorted(values)
 
-    with directory.joinpath("variants.json").open("w") as f:
+    with (directory / "variants.json").open(mode="w") as f:
```

Comment:
I personally prefer `joinpath()` over adding extra parentheses but that's just a matter of taste.